Fix: Mudlet freeze when receiving large amounts of data

mpconley opened the pull request:
#### Brief overview of PR changes/additions

Fixes a freeze that could occur when receiving exactly 100,000 bytes of data from the game server at once.

#### Motivation for adding to Mudlet

Prevents Mudlet from becoming unresponsive and consuming excessive memory when games send large amounts of data (e.g., many clickable links or rich content).

#### Other info (issues closed, discussion etc)

Closes #8643

Discovered while investigating OSC 8 hyperlink performance in PR #8623.

mpconley committed:
initial

mpconley commented:
There is a bug here that freezes Mudlet. Another way to solve it would remove this loop entirely, but it would be a bigger diff.

I am working my way around this and send smaller batches of data, but this seems like a legit find as it eats up all the memory when it does get triggered.

vadi2 reviewed (commented):
This is a demonstration in why I don't like while loops. 

As this is a pretty rare scenario, do you mind giving it a go to refactor the while loop a away?

mpconley committed:
Merge branch 'development' into fix/infinite-loop-8643

mpconley committed:
Fix infinite loop in processSocketData and optimize string allocation

- Remove do-while loop that caused infinite loop when BUFFER_SIZE bytes were read
- The loop body never reads new data; Qt invokes socketReadyToBeRead() again for more data
- Add std::string preallocation (BUFFER_SIZE * 4) for decompressed data to reduce reallocations
- Keep MAIN_LOOP_END label for proper GA signal handling via goto in MCCP detection path
- Simplify comment explaining the single-pass processing design

mpconley commented:
Done and this is working fine. The change is removed the loop and added some whitespace around the blocks as-is tradition. The TODOs from DangerJS were in there already.

SlySven commented:
> This is a demonstration in why I don't like while loops.

Strangely this wasn't a while-loop but a do-loop one and the semantics are significant - as I was taught when I did my City & Guilds Advance diploma in the C programming language back in the 90s :grinning:. The important different is that the do-loop ALWAYS makes one pass through the loop and then does the test whether to go around again - whereas a while loop does the test before any passes are made and may skip the loop entirely.

This do-loop was put in by @ahmedcharles in 07ff975393f3b0992b9b08d88cb3f50c7d1f516d (in 2014) - so are you saying that we do not need it even to solve the problem that it was fixing?

mpconley commented:
@SlySven You raise an excellent point about the semantic difference between `do-while` vs `while` loops, and your question about whether we still need the loop to solve the original problem is spot-on.

I investigated Ahmed's 2014 commit (07ff975) that introduced the do-while loop. The **original intention was sound** - it was designed to handle MCCP buffer management scenarios where exactly BUFFER_SIZE bytes might be processed (especially after decompression), and more data could be immediately available that should be processed in the same function call.

However, there was a **critical flaw in the implementation**: the loop body never actually reads new data from the socket. The `amount` parameter is fixed for the entire function call - it's the data that was already read by `handle_socket_signal_readyRead()`. So when `datalen == BUFFER_SIZE`, the loop just reprocesses the same data infinitely.

**Why removing the loop entirely is safe:**

1. **Qt's signal/slot mechanism is reliable** - When more data arrives on the socket, Qt will emit `readyRead()` again, triggering a fresh call to `processSocketData()` with new data.

2. **MCCP scenarios are properly handled** - The decompression happens within the single pass. If there's more compressed data waiting on the socket, Qt will signal `readyRead()` again.

3. **Ahmed's fix would have needed to actually read from the socket** - Something like:
   ```cpp
   } while (datalen == BUFFER_SIZE && mpSocket->bytesAvailable() > 0);
   ```
   But this adds complexity and potential edge cases that Qt's event loop already handles reliably.

The current approach is simpler, safer, and maintains the same functional behavior without the infinite loop risk. Ahmed's original problem (MCCP buffer management) is still solved, just through Qt's event mechanism rather than manual looping.

SlySven commented on src/ctelnet.cpp:
```diff
@@ -4642,203 +4642,219 @@ void cTelnet::processSocketData(char* in_buffer, int amount, const bool loopback
-                    cleandata.push_back('\n');
                 }
+
+                cleandata.push_back('\xff');
```
\<off-topic>*I've still got to come up with a way around this kludge - whilst it was (for the original ASCII only encoding handling) and is (for UTF-8) okay to use this byte value to flag the reception of a GA/EOR - it is broken for all encodings that DO use the `\xff` byte value - and there are quite a few...*\</off-topic>

SlySven reviewed (commented):
It looks good to me,  but I haven't actually run it to test it...

SlySven commented:
Um, @mpconley - have you been replaced by an AI? :worried: That detailed, well spoken, analysis doesn't sound entirely *human*! :grinning: Congratulations on finding an implementation bug that has been around for **11 years** :astonished:

mpconley commented:
> Um, @mpconley - have you been replaced by an AI? 😟 That detailed, well spoken, analysis doesn't sound entirely _human_! 😀 Congratulations on finding an implementation bug that has been around for **11 years** 😲

Yes, sometimes I will make AI respond straight from GitHub MCP Server. Human is still in the loop!
